### DISCUSSION TITLE
How to exclude imported dependencies in stories from being built into package?

### DISCUSSION THREAD

**Discussion by cforsyth68** [nrwl/nx#17168] [General]
I have an NX repo set up to build individual components of a Design System, in their own library packages.

I have a /stories folder in each project, under /libs/<component>/src/lib folder.

I use other components within the stories to make the stories look nicer and as an example of using my own design system.

These are imported with import statements within the stories.

When building the component, these imports get added as peerDependencies, but I do not need them as peerDeps as these are only used by stories, which is actually set up to run in an app of the same nx repo. I loosely followed info written here: https://nx.dev/recipes/storybook/one-storybook-for-all to achieve this.

I did find the "implicitDependencies" option of project.json file, but this would mean each component would have it's own manually configured list of projects to ignore. (using the syntax ["!<project_name>"]).  This does work but it's too much manual effort.

Is there maybe another configuration option somewhere I can use that allows me to exclude this /stories folder from the build for all my libraries?

**Comment by kevindice** [comment]
Hi @cforsyth68 - I'm currently doing some work + research in the area of Nx's dependency graph.  I see that there is a little bit of motion within the source code for dependency graph construction since 16.2.1 was released.

What I have in mind is a way to classify individual files as either "runtime" or "development".  This allows dependency edges to then be classified as "runtime" or "development" as well.  Then downstream plugins and other custom tasks may use this information as they see fit to generated appropriate dependency lists based on the given use case.

When being analyzed, we would make the "runtime" vs "development" classification based on the file name.

For my noodling around, I've got this crude function defined, but I envision that this would be a configurable option and specified w/ glob patterns:

```
const isRuntime = (filename: string) => {
  if (
    filename.match(/\.spec\.ts/) ||
    filename.match(/(test-setup)\.ts/) ||
    filename.match('__(mock|mocks)__') ||
    filename.match('metro.config.js') ||
    filename.match('environment.js')
  ) {
    return false;
  }

  return true;
};
```
...basically, assume all files are runtime files unless they match a provided pattern.

For context, I am working on an nx-powered repository with multiple react-native apps and several dozen modules.  Within this repo I too have a collection of design system modules and, at some point, may want to publish these to other teams within my organization, at which point, I would want a correct and auto-generated dependency list.

And, like you, I've chosen to co-locate my stories alongside the component implementation within each module.

So - I think we've got similar problem to solve here.

At present, the initial construction of the dependency graph takes place within a plugin named `nx-js-graph-plugin` which is currently hard-coded as the 1st plugin to run:  https://github.com/nrwl/nx/blob/16.2.1/packages/nx/src/utils/nx-plugin.ts#L166-L176

Further graph processing may then happen on any plugins which export a `processProjectGraph` method:  https://github.com/nrwl/nx/blob/16.2.1/packages/nx/src/project-graph/build-project-graph.ts#L233-L244

Corresponding documentation on how to create a plugin lives here:  https://nx.dev/plugins/recipes/project-graph-plugins#implementing-a-project-graph-processor

---

I would argue that the solution to our problems has two parts:
- We need to update the `nx-js-graph-plugin` to have first-party support for classifying files and their resulting dependencies as "runtime" or "development"
- We need to update our downstream integrations to make use of this newly provided information to achieve our desired outcomes.

The dependency graph is used for many purposes, and I don't think it would be appropriate to make destructive edits to it downstream of the `nx-js-graph-plugin`.  The most convenient time to annotate files + dependencies is at the time of graph construction.  By the time we access `graph.externalDependencies['my-module-name']` and see an object containing

```
  'npm:@storybook/addon-knobs': {
    type: 'npm',
    name: 'npm:@storybook/addon-knobs',
    data: {
      version: '^5.3',
      packageName: '@storybook/addon-knobs',
      hash: 'e5be50134a232b76037d7e5c16bb1f2b3d03db6d7d17f8916331e991ec063e5a'
    }
  },

```

the information as to which file imported this has already been lost.

And further, I'd argue that a "runtime" vs "development" distinction is broadly useful and is worth having embedded in the dependency graph for a multitude of uses.

For my purposes w/ React Native development, I've grown a bit tired of seeing `@testing-library/react-native` show up in my app's package.json.  It certainly does no harm, but I am concerned about my overall path forward when publishing some of my repo's modules for use by another team within my org.  I don't want to be shipping a `package.json` that declares `@storybook/addon-ondevice-actions` as a runtime dependency when it is not.

I'm gradually increasing my understanding of the dependency graph construction and may produce a PR here soon.  Would love to colab on that if it interests you.

**Reply by kevindice** [reply]
One difference in our setups is that I have explicit imports from my storybook app to all of my stories, and this is maintained by a git hook running this script:

```
const { glob } = require('glob');
const fs = require('fs');
const path = require('path');
const { createHash } = require('crypto')

const stories = glob.globSync('./libs/**/*.stories.tsx');

const data = `export const loadCollectedStories = () => {
${stories.sort().map(story => {
  const buff = fs.readFileSync(story);
  const hash = createHash("md5").update(buff).digest("hex");
  const relativePath = path.relative('./apps/storybook/src/storybook', story);
  return `  /*\n   * story: ${path.basename(story)}\n   * md5: ${hash}\n   */\n  require('${relativePath}');\n`
}).join('\n')}};
`
fs.writeFileSync('./apps/storybook/src/storybook/auto-collected-stories.js', data, 'utf8');

```

<details><summary>Resulting file (w/ slightly wonky formatting thanks to GitHub)</summary>
<pre><code>
export const loadCollectedStories = () => {
  /*
     * story: accordion.stories.tsx
     * md5: 430c65f6fd80ff8ff746f1a0c420d4b5
     */
  require('../../../../libs/design-system/accordion/src/accordion.stories.tsx');

  /*
     * story: Alert.stories.tsx
     * md5: 47493f55e2df990f170f72f54da374e4
     */
  require('../../../../libs/design-system/alert/src/Alert.stories.tsx');

  /*
     * story: BackButton.stories.tsx
     * md5: f9d676785dc7abc08d2b95016605469b
     */
  require('../../../../libs/design-system/back-button/src/BackButton.stories.tsx');

  /*
     * story: HeroImage.stories.tsx
     * md5: 8dcddaeea9504107bb0dd1fbce37ada2
    */
  require('../../../../libs/design-system/hero-img/src/HeroImage.stories.tsx');
};
</code></pre>
</details>
